src/object/srv_obj.c

```diff
@@ -791,58 +734,235 @@ ds_obj_check_dtx_config(uint32_t flags, struct dtx_id *dti)
 	return 0;
 }
 
+/* Various check before access VOS */
+static int
+ds_pre_check(daos_unit_oid_t oid, uint32_t rpc_map_ver, uuid_t hdl_uuid,
+	     uuid_t co_uuid, uint32_t opc, uint32_t flags, struct dtx_id *dti,
+	     struct ds_cont_hdl **hdlp, struct ds_cont_child **contp)
+{
+	uint32_t	map_ver;
+	int		rc;
+
+	*hdlp = NULL;
+	*contp = NULL;
+	rc = ds_check_container(hdl_uuid, co_uuid, opc, hdlp, contp);
+	if (rc)
+		return rc;
+
+	D_ASSERT((*hdlp)->sch_pool != NULL);
+	map_ver = (*hdlp)->sch_pool->spc_map_version;
+
+	if (rpc_map_ver < map_ver) {
+		D_DEBUG(DB_IO, "stale version req %d map_version %d\n",
+			rpc_map_ver, map_ver);
+		if (obj_modify_rpc(opc))
+			D_GOTO(out_put, rc = -DER_STALE);
+		/* It is harmless if fetch with old pool map version. */
```

Comment:
Why? I think it was a bug. No mater fetch or update, we have to make sure the pool map version is uptodate, no?

Reply:
oh, for fetch I think we can go ahead at least for current replication implementation.  I will check later for EC.  Though it can do better job like also trigger the pool map update on client for this case.

Reply:
This isn't a problem of this patch, but we need to consider that if a pool target (VOS) is DOWN, the VOS xstream can still serve RPCs.

We may need to fix it in other patch (either make sure the xstream not serving RPC when the target is DOWN or check pool map version for both fetch and update?)

Reply:
@Nasf-Fan , when a target is marked as DOWN in the pool map for whatever reason (SSD device is detected as faulty or the server node is detected as dead by SWIM),  we don't guarantee this DOWN target stop serving update/fetch requests, so stale (incorrect) data could be return to client here.

I think that's the same reason we check pool map version for update, right? (I just don't quite see why we treat update and fetch differently here)

Reply:
What's the relationship between target (marked as) DOWN and old pool map version? If the DOWN target still can offer IO service, why its pool version cannot be refreshed?

Reply:
currently as that there is impossible that partial vos targets failed on a node that with new pool map version created but this node and its other vos targets still allive (i.e. currently as long as the node alive all its vos tgts are alive). So seems fine?